<a href="https://colab.research.google.com/github/heeeee-github/Stock_Analysis_in_Python_230327/blob/main/2%EC%A3%BC%EC%B0%A8/pandas_exercise_skeleton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [31]:
df = pd.read_csv('https://raw.githubusercontent.com/zzhining/ml_basic/main/dataset/insurance.csv', index_col= False)
#df = pd.read_csv('insurance.csv', index_col= False)

df.head(5)

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [4]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [32]:
df['double_bmi'] = df['bmi'] *2

df.head(5)

age     sex     bmi  children smoker     region      charges  double_bmi
0   19  female  27.900         0    yes  southwest  16884.92400       55.80
1   18    male  33.770         1     no  southeast   1725.55230       67.54
2   28    male  33.000         3     no  southeast   4449.46200       66.00
3   33    male  22.705         0     no  northwest  21984.47061       45.41
4   32    male  28.880         0     no  northwest   3866.85520       57.76

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [33]:
df['debt'] = 0
df.head(5)

age     sex     bmi  children smoker     region      charges  double_bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     0  
2     0  
3     0  
4     0

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [34]:
for i in range(len(df['debt'])):
    df['debt'][i] = i  # 0부터 값이 증가하도록 채우기

df.head()


<ipython-input-34-eaf2ee5303aa>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['debt'][i] = i  # 0부터 값이 증가하도록 채우기


age     sex     bmi  children smoker     region      charges  double_bmi  \
0   19  female  27.900         0    yes  southwest  16884.92400       55.80   
1   18    male  33.770         1     no  southeast   1725.55230       67.54   
2   28    male  33.000         3     no  southeast   4449.46200       66.00   
3   33    male  22.705         0     no  northwest  21984.47061       45.41   
4   32    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [10]:
df.iloc[1]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double_bmi        67.54
debt                  0
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [ ]:
df.drop(['double_bmi'], axis=1)

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [18]:
df[['age','sex']]

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [35]:
for i in range(len(df['age'])):
    df['age'][i] = df['age'][i] + 1 # 0부터 값이 증가하도록 채우기

df.head()


<ipython-input-35-a5f3e22244e8>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['age'][i] = df['age'][i] + 1 # 0부터 값이 증가하도록 채우기


age     sex     bmi  children smoker     region      charges  double_bmi  \
0   20  female  27.900         0    yes  southwest  16884.92400       55.80   
1   19    male  33.770         1     no  southeast   1725.55230       67.54   
2   29    male  33.000         3     no  southeast   4449.46200       66.00   
3   34    male  22.705         0     no  northwest  21984.47061       45.41   
4   33    male  28.880         0     no  northwest   3866.85520       57.76   

   debt  
0     0  
1     1  
2     2  
3     3  
4     4

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [ ]:
######## 미작성

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [36]:
df.sort_index(axis=1)

age     bmi      charges  children  debt  double_bmi     region     sex  \
0      20  27.900  16884.92400         0     0       55.80  southwest  female   
1      19  33.770   1725.55230         1     1       67.54  southeast    male   
2      29  33.000   4449.46200         3     2       66.00  southeast    male   
3      34  22.705  21984.47061         0     3       45.41  northwest    male   
4      33  28.880   3866.85520         0     4       57.76  northwest    male   
...   ...     ...          ...       ...   ...         ...        ...     ...   
1333   51  30.970  10600.54830         3  1333       61.94  northwest    male   
1334   19  31.920   2205.98080         0  1334       63.84  northeast  female   
1335   19  36.850   1629.83350         0  1335       73.70  southeast  female   
1336   22  25.800   2007.94500         0  1336       51.60  southwest  female   
1337   62  29.070  29141.36030         0  1337       58.14  northwest  female   

     smoker  
0       yes  
1        no  
2        no  
3        no  
4        no  
...     ...  
1333     no  
1334     no  
1335     no  
1336     no  
1337    yes  

[1338 rows x 9 columns]

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [12]:
df.sort_values(by = ['age'])

age     sex     bmi  children smoker     region      charges  \
1248   18  female  39.820         0     no  southeast   1633.96180   
482    18  female  31.350         0     no  southeast   1622.18850   
492    18  female  25.080         0     no  northeast   2196.47320   
525    18  female  33.880         0     no  southeast  11482.63485   
529    18    male  25.460         0     no  northeast   1708.00140   
...   ...     ...     ...       ...    ...        ...          ...   
398    64    male  25.600         2     no  southwest  14988.43200   
335    64    male  34.500         0     no  southwest  13822.80300   
378    64  female  30.115         3     no  northwest  16455.70785   
1265   64    male  23.760         0    yes  southeast  26926.51440   
635    64    male  38.190         0     no  northeast  14410.93210   

      double_bmi  debt  
1248       79.64     0  
482        62.70     0  
492        50.16     0  
525        67.76     0  
529        50.92     0  
...          ...   ...  
398        51.20     0  
335        69.00     0  
378        60.23     0  
1265       47.52     0  
635        76.38     0  

[1338 rows x 9 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [13]:
df['age'].mean()

39.20702541106129

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

In [37]:
df.cov()

age           bmi    children       charges  \
age           197.401387      9.362337    0.719303  5.087480e+04   
bmi             9.362337     37.187884    0.093795  1.464730e+04   
children        0.719303      0.093795    1.453213  9.926742e+02   
charges     50874.802298  14647.304426  992.674197  1.466524e+08   
double_bmi     18.724673     74.375767    0.187590  2.929461e+04   
debt         -170.829095    -85.224706   11.747569 -1.578122e+04   

              double_bmi           debt  
age            18.724673    -170.829095  
bmi            74.375767     -85.224706  
children        0.187590      11.747569  
charges     29294.608853  -15781.223645  
double_bmi    148.751534    -170.449413  
debt         -170.449413  149298.500000

In [38]:
df.corr()

age       bmi  children   charges  double_bmi      debt
age         1.000000  0.109272  0.042469  0.299008    0.109272 -0.031467
bmi         0.109272  1.000000  0.012759  0.198341    1.000000 -0.036169
children    0.042469  0.012759  1.000000  0.067998    0.012759  0.025221
charges     0.299008  0.198341  0.067998  1.000000    0.198341 -0.003373
double_bmi  0.109272  1.000000  0.012759  0.198341    1.000000 -0.036169
debt       -0.031467 -0.036169  0.025221 -0.003373   -0.036169  1.000000

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>